In [2]:
#libarary
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib
matplotlib.style.use('ggplot')
%matplotlib inline

## Q1: PCA for Reduced Dimensionality in Clustering

[Dataset: segmentation_data.zip]

For this problem you will use an image segmentation data set for clustering. You will experiment with using PCA as an approach to reduce dimensionality and noise in the data. You will compare the results of clustering the data with and without PCA using the provided image class assignments as the ground truth. The data set is divided into three files. The file "segmentation_data.txt" contains data about images with each line corresponding to one image. Each image is represented by 19 features (these are the columns in the data and correspond to the feature names in the file "segmentation_names.txt". The file "segmentation_classes.txt" contains the class labels (the type of image) and a numeric class label for each of the corresponding images in the data file. After clustering the image data, you will use the class labels to measure completeness and homogeneity of the generated clusters. The data set used in this problem is based on the Image Segmentation data set at the UCI Machine Learning Repository.



## 1a.

Load in the image data matrix (with rows as images and columns as features). Also load in the numeric class labels from the segmentation class file. Using your favorite method (e.g., sklearn's min-max scaler), perform min-max normalization on the data matrix so that each feature is scaled to [0,1] range.

### Load in the image data matrix

In [3]:
image_matrix = pd.read_table('segmentation_data/segmentation_data.txt',na_values='?',header= None,sep=',')
image_matrix.head()

0      1   2    3    4         5         6         7         8   \
0  110.0  189.0   9  0.0  0.0  1.000000  0.666667  1.222222  1.186342   
1   86.0  187.0   9  0.0  0.0  1.111111  0.720082  1.444444  0.750309   
2  225.0  244.0   9  0.0  0.0  3.388889  2.195113  3.000000  1.520234   
3   47.0  232.0   9  0.0  0.0  1.277778  1.254621  1.000000  0.894427   
4   97.0  186.0   9  0.0  0.0  1.166667  0.691215  1.166667  1.005540   

          9          10         11         12        13         14         15  \
0  12.925926  10.888889   9.222222  18.666668 -6.111111 -11.111111  17.222221   
1  13.740741  11.666667  10.333334  19.222221 -6.222222 -10.222222  16.444445   
2  12.259259  10.333334   9.333334  17.111110 -5.777778  -8.777778  14.555555   
3  12.703704  11.000000   9.000000  18.111110 -5.111111 -11.111111  16.222221   
4  15.592592  13.888889  11.777778  21.111110 -5.111111 -11.444445  16.555555   

          16        17        18  
0  18.666668  0.508139  1.910864  
1  19.222221  0.463329  1.941465  
2  17.111110  0.480149  1.987902  
3  18.111110  0.500966  1.875362  
4  21.111110  0.442661  1.863654

In [4]:
#check for missing values
print "attribute with missing values: \n"
print [col for col in image_matrix.columns if image_matrix[col].isnull().any()]

attribute with missing values: 

[]


In [5]:
image_matrix.shape

(2100, 19)

### load in the numeric class labels (type of image)

In [6]:
image_class = pd.read_table('segmentation_data/segmentation_classes.txt',sep='\t',header=None,na_values='?')
image_class.head()

0  1
0  GRASS  0
1  GRASS  0
2  GRASS  0
3  GRASS  0
4  GRASS  0

In [7]:
#check for missing values
print "attribute with missing values: \n"
print [col for col in image_class.columns if image_class[col].isnull().any()]

attribute with missing values: 

[]


In [8]:
image_class.shape

(2100, 2)

### min-max normalization on the data matrix 

In [9]:
from sklearn.preprocessing import MinMaxScaler

In [10]:
min_max_scaler = MinMaxScaler()

In [11]:
image_matrix_norm = min_max_scaler.fit_transform(image_matrix)
image_matrix_norm

array([[ 0.43083004,  0.74166667,  0.        , ...,  0.12371135,
         0.50813884,  0.83184923],
       [ 0.33596838,  0.73333333,  0.        , ...,  0.12739322,
         0.46332908,  0.83698646],
       [ 0.88537549,  0.97083333,  0.        , ...,  0.11340205,
         0.48014903,  0.84478233],
       ..., 
       [ 0.50197628,  0.625     ,  0.        , ...,  0.07216495,
         0.5409177 ,  0.17591546],
       [ 0.58893281,  0.6125    ,  0.        , ...,  0.08100147,
         0.50308645,  0.18478933],
       [ 0.48616601,  0.62916667,  0.        , ...,  0.09646539,
         0.4799313 ,  0.17037463]])

## 1b.

Next, Perform Kmeans clustering (for this problem, use the Kmeans implementation in scikit-learn) on the image data (since there are a total 7 pre-assigned image classes, you should use K = 7 in your clustering). Use Euclidean distance as your distance measure for the clustering. Print the cluster centroids (use some formatting so that they are visually understandable). Compare your 7 clusters to the 7 pre-assigned classes by computing the Completeness and Homogeneity values of the generated clusters.

### Kmeans clustering (k=7,Euclidean distance)

In [12]:
from sklearn import cluster
centroids,pred_Cluster,inertia = cluster.k_means(image_matrix_norm,n_clusters=7)

### Print the cluster centroids

In [13]:
image_name= pd.read_table('segmentation_data/segmentation_names.txt',header=None,na_values='?')
image_name.head()

0
0   REGION-CENTROID-COL
1   REGION-CENTROID-ROW
2    REGION-PIXEL-COUNT
3  SHORT-LINE-DENSITY-5
4  SHORT-LINE-DENSITY-2

In [14]:
colname=["Centroid 1","Centroid 2","Centroid 3","Centroid 4","Centroid 5","Centroid 6","Centroid 7"]

In [15]:
centroidsDF= pd.DataFrame(centroids.T,columns=colname,index=image_name[0].tolist())

In [16]:
print '                                  Cluster Centroids','\n'
centroidsDF

                                  Cluster Centroids 



Centroid 1  Centroid 2  Centroid 3  Centroid 4  \
REGION-CENTROID-COL     0.750696    0.770674    0.513994    0.535099   
REGION-CENTROID-ROW     0.534564    0.425215    0.808937    0.150167   
REGION-PIXEL-COUNT      0.000000    0.000000    0.000000    0.000000   
SHORT-LINE-DENSITY-5    0.040000    0.013978    0.077441    0.027778   
SHORT-LINE-DENSITY-2    0.038462    0.022581    0.005051    0.001667   
VEDGE-MEAN              0.114419    0.040237    0.054474    0.030228   
VEDGE-SD                0.019301    0.002989    0.001407    0.000543   
HEDGE-MEAN              0.109240    0.023122    0.046335    0.026766   
HEDGE-SD                0.017999    0.002088    0.001401    0.000587   
INTENSITY-MEAN          0.300955    0.041138    0.108790    0.823246   
RAWRED-MEAN             0.279908    0.035005    0.091403    0.779716   
RAWBLUE-MEAN            0.352738    0.058340    0.092414    0.894170   
RAWGREEN-MEAN           0.265920    0.028766    0.142676    0.788761   
EXRED-MEAN              0.592572    0.778473    0.679161    0.270665   
EXBLUE-MEAN             0.451124    0.223939    0.079002    0.666373   
EXGREEN-MEAN            0.309158    0.486509    0.821287    0.289386   
VALUE-MEAN              0.352822    0.059295    0.134901    0.894170   
SATURATION-MEAN         0.302342    0.538181    0.414491    0.211804   
HUE-MEAN                0.164764    0.243969    0.892333    0.125066   

                      Centroid 5  Centroid 6  Centroid 7  
REGION-CENTROID-COL     0.254169    0.256103    0.302506  
REGION-CENTROID-ROW     0.459974    0.393468    0.530862  
REGION-PIXEL-COUNT      0.000000    0.000000    0.000000  
SHORT-LINE-DENSITY-5    0.026256    0.074510    0.052260  
SHORT-LINE-DENSITY-2    0.013699    0.019118    0.046610  
VEDGE-MEAN              0.037274    0.077343    0.100817  
VEDGE-SD                0.002364    0.004100    0.009420  
HEDGE-MEAN              0.027874    0.060574    0.083972  
HEDGE-SD                0.002017    0.004967    0.011043  
INTENSITY-MEAN          0.026013    0.148187    0.400608  
RAWRED-MEAN             0.017925    0.138085    0.370347  
RAWBLUE-MEAN            0.042306    0.185160    0.472461  
RAWGREEN-MEAN           0.016504    0.118539    0.353036  
EXRED-MEAN              0.769800    0.716906    0.497146  
EXBLUE-MEAN             0.216011    0.344014    0.570882  
EXGREEN-MEAN            0.508117    0.355458    0.213054  
VALUE-MEAN              0.043295    0.185413    0.472461  
SATURATION-MEAN         0.801687    0.411986    0.302263  
HUE-MEAN                0.181153    0.201890    0.163879

### Completeness and Homogeneity values 

In [17]:
from sklearn.metrics import completeness_score, homogeneity_score

In [18]:
print 'completeness_score = ', completeness_score(image_class[1],pred_Cluster)
print 'homogeneity_score = ', homogeneity_score(image_class[1],pred_Cluster)

completeness_score =  0.613187012485
homogeneity_score =  0.611502116337


## 1c.

Perform PCA on the normalized image data matrix. You may use the linear algebra package in Numpy or the Decomposition module in scikit-learn (the latter is much more efficient). Analyze the principal components to determine the number, r, of PCs needed to capture at least 95% of variance in the data. Then use these r components as features to transform the data into a reduced dimension space. [See the PCA Clustering Notebook from class for an example of how these steps are performed.]

### PCA

In [19]:
from sklearn import decomposition

#### determine the number, r, of PCs

In [20]:
pca = decomposition.PCA(n_components=19)
pca.fit(image_matrix_norm)
print pca.explained_variance_ratio_.cumsum()

[ 0.60714234  0.73911213  0.84034986  0.88578525  0.92125886  0.94113922
  0.96005892  0.97621301  0.98686902  0.99398239  0.99790443  0.99947943
  0.99996858  1.          1.          1.          1.          1.          1.        ]


7 PCs needed to capture at least 95% of variation in the data

#### Transform the data into a reduced dimension space( 7  components)

In [21]:
pca = decomposition.PCA(n_components=7)
np.set_printoptions(precision=2,suppress=True)
pcs=pca.fit(image_matrix_norm).transform(image_matrix_norm)

In [22]:
print pcs

[[-0.69  0.53  0.25 ..., -0.08  0.05 -0.05]
 [-0.67  0.51  0.34 ..., -0.04  0.06 -0.04]
 [-0.71  0.77 -0.16 ..., -0.17  0.04 -0.06]
 ..., 
 [-0.51 -0.13 -0.08 ..., -0.03 -0.03 -0.11]
 [-0.48 -0.09 -0.16 ...,  0.   -0.   -0.09]
 [-0.44 -0.11 -0.05 ...,  0.02 -0.21  0.15]]


## 1d.

Perform Kmeans again, but this time on the lower dimensional transformed data. Then, compute the Completeness and Homogeneity values of the new clusters.

### Kmean

In [23]:
from sklearn import cluster
centroids,pred_Cluster,inertia = cluster.k_means(pcs,n_clusters=7)

In [24]:
pcs_name=['PC 1','PC 2','PC 3','PC 4', 'PC 5', 'PC 6', 'PC 7']
colname=["Centroid 1","Centroid 2","Centroid 3","Centroid 4","Centroid 5","Centroid 6","Centroid 7"]
print '                         Cluster Centroids','\n'
pd.DataFrame(centroids.T, index=pcs_name, columns=colname)

                         Cluster Centroids 



Centroid 1  Centroid 2  Centroid 3  Centroid 4  Centroid 5  Centroid 6  \
PC 1   -0.206229   -0.619268   -0.603705    0.176404    1.414527   -0.511579   
PC 2   -0.246205    0.640250   -0.355503    0.043698    0.087223   -0.064908   
PC 3    0.152785    0.195829    0.109197   -0.265370    0.036765   -0.336145   
PC 4    0.056522   -0.086856   -0.129799    0.184124   -0.173195   -0.065365   
PC 5    0.130585   -0.067760   -0.130911    0.027075   -0.029922    0.078809   
PC 6   -0.005540    0.008866   -0.021603    0.024341   -0.008973    0.006240   
PC 7    0.032849    0.038372   -0.043882    0.003264   -0.021573   -0.026261   

      Centroid 7  
PC 1    0.436890  
PC 2   -0.104936  
PC 3    0.164951  
PC 4    0.234379  
PC 5   -0.046159  
PC 6   -0.007512  
PC 7    0.015253

### Completeness and Homogeneity values 

In [25]:
from sklearn.metrics import completeness_score, homogeneity_score
print 'completeness_score = ', completeness_score(image_class[1],pred_Cluster)
print 'homogeneity_score = ', homogeneity_score(image_class[1],pred_Cluster)

completeness_score =  0.610795506369
homogeneity_score =  0.609136404973


## 1e.

Discuss your observations based on the comparison of the two clustering results.

__Normalized__

completeness_score =  0.613187012485

homogeneity_score =  0.611502116337

__Normalized and Reduced(7 components)__

completeness_score =  0.611574926992

homogeneity_score =  0.609973609217

__Discussion:__

we can see that clusteing results for both are similar. In this case, the  Normalized and Reduced(7 components) data is preffered since we can get almost the same result by using smaller set of the data

## Q2: Association Rule Discovery

[Dataset: playlists.zip]

For this problem you will experiment with association rule mining using Apriori algorithm discussed in class. You will use a modified version of the Apriori implementation in Machine Learning in Action (it has been modified to compute lift values for rules in addition to confidence). [See Associations Notebook from class for an example of using this module.]

The data set you will use is based on a music playlist data set obtained from Yes.com. [See the full description of this data]. We will only use a portion of this data. The provided data archive contains two files. The file "playlists.txt" contains on each line a sequence of songs played as part of one playlist. The songs are represented by integer values. The file "song_names.txt" contains the mapping between the integer codes and song tiles and artists (format of the song names is [song title]::[artist]). You will need both of these files to generate association rules.

## 2a.

Load the playlist data into a Python nested list, and the song_names data into an appropriate data structure.

### playlist data

In [26]:
playlist = []
for line in open('playlists/playlists.txt'):
    x = [int(i) for i in line.split()]
    playlist.append(x)
    
playlist[0:10]

[[0, 1, 2, 3, 4, 5, 6, 7, 8],
 [9, 10, 11],
 [12, 13, 14, 15],
 [16, 17, 18],
 [19],
 [2],
 [20, 21, 22, 15, 23, 8, 24, 25],
 [26],
 [27, 18],
 [28, 29]]

### song_names data

In [27]:
song_names = np.genfromtxt("playlists/song_names.txt", delimiter="\t", usecols=1, dtype=None)
song_names

array(['Gucci-Time::Gucci-Mane', 'Aston-Martin-Music::Rick-Ross',
       'Get-Back-Up::T.I.', ...,
       'All-I-Want-For-Christmas-Is-You::Mariah-Carey',
       'Jingle-Bell-Rock::Daryl-Hall-&-John-Oates',
       'Do-You-Hear-What-I-Hear?::Whitney-Houston'], 
      dtype='|S86')

In [38]:
item_meaning = dict((i, song_names[i]) for i in range(len(song_names)))

## 2b.

Run Apriori on the playlist data using a min-support value of 0.002.

In [30]:
import apriori2 as ap
Lift, support = ap.apriori(playlist, 0.002)

In [31]:
Lift

[[frozenset({1938}),
  frozenset({2776}),
  frozenset({2508}),
  frozenset({2792}),
  frozenset({1250}),
  frozenset({2841}),
  frozenset({3019}),
  frozenset({1769}),
  frozenset({195}),
  frozenset({546}),
  frozenset({3046}),
  frozenset({3094}),
  frozenset({2116}),
  frozenset({1448}),
  frozenset({2117}),
  frozenset({2057}),
  frozenset({405}),
  frozenset({490}),
  frozenset({786}),
  frozenset({1616}),
  frozenset({1749}),
  frozenset({2836}),
  frozenset({29}),
  frozenset({2747}),
  frozenset({1057}),
  frozenset({78}),
  frozenset({122}),
  frozenset({175}),
  frozenset({2955}),
  frozenset({1996}),
  frozenset({232}),
  frozenset({2826}),
  frozenset({372}),
  frozenset({3007}),
  frozenset({2966}),
  frozenset({807}),
  frozenset({2267}),
  frozenset({2229}),
  frozenset({2762}),
  frozenset({2680}),
  frozenset({2417}),
  frozenset({2923}),
  frozenset({26}),
  frozenset({196}),
  frozenset({636}),
  frozenset({701}),
  frozenset({3052}),
  frozenset({2998}),
  frozenset

In [32]:
support

{frozenset({241, 330}): 9.643201542912247e-05,
 frozenset({215, 2776}): 0.00012054001928640309,
 frozenset({2609}): 0.0024108003857280617,
 frozenset({511, 565}): 0.0001687560270009643,
 frozenset({1397, 2966}): 4.8216007714561236e-05,
 frozenset({552, 637}): 0.00024108003857280618,
 frozenset({2379}): 0.0005785920925747348,
 frozenset({565, 2917}): 0.00012054001928640309,
 frozenset({1988, 2037}): 0.0002651880424300868,
 frozenset({2790, 2866, 2924}): 0.0011089681774349085,
 frozenset({78, 2572}): 9.643201542912247e-05,
 frozenset({2415, 2940}): 0.00019286403085824494,
 frozenset({520, 2417}): 0.00019286403085824494,
 frozenset({319}): 0.002627772420443587,
 frozenset({1250, 2552}): 2.4108003857280618e-05,
 frozenset({2020, 2047}): 0.0001687560270009643,
 frozenset({285, 2584}): 0.0003857280617164899,
 frozenset({2087, 2417}): 2.4108003857280618e-05,
 frozenset({405, 2812}): 7.232401157184185e-05,
 frozenset({2671}): 0.0005303760848601736,
 frozenset({796}): 0.0009884281581485053,
 fr

## 2c.

Generate rules from the frequent itemsets discovered in the previous step (you can try different metrics ('lift' or 'confidence' to see which gives you more useful and interesting results). At minimum, you should generate rules with a minimum lift value of 20.0. If you use confidence, you may want to set the min-confidence threshold to about 0.5.

### metrics: confidence

In [34]:
ruleList_con = ap.generateRules(Lift, support, metric='confidence', minMetric=0.5)

frozenset([2840]) --> frozenset([2839]) conf: 0.511727078891  lift: 29.1571967478
frozenset([2883]) --> frozenset([2878]) conf: 0.528138528139  lift: 49.9024741394
frozenset([2826]) --> frozenset([2839]) conf: 0.617511520737  lift: 35.1845850003
frozenset([2774]) --> frozenset([2018]) conf: 0.514925373134  lift: 34.843563585
frozenset([2878]) --> frozenset([2839]) conf: 0.507972665148  lift: 28.943277679
frozenset([2878]) --> frozenset([5]) conf: 0.535307517084  lift: 32.7018494973
frozenset([44]) --> frozenset([2018]) conf: 0.502538071066  lift: 34.0053494092
frozenset([2871]) --> frozenset([5]) conf: 0.535294117647  lift: 32.7010309278
frozenset([5]) --> frozenset([2839]) conf: 0.511045655376  lift: 29.1183705838
frozenset([2871]) --> frozenset([2839]) conf: 0.505882352941  lift: 28.8241758242
frozenset([2883]) --> frozenset([5]) conf: 0.510822510823  lift: 31.2060644314
frozenset([2512]) --> frozenset([20]) conf: 0.524038461538  lift: 33.2371794872
frozenset([3018]) --> frozenset([3

In [39]:
ap.pntRules(ruleList_con, item_meaning)

All-Of-The-Lights::Kanye-West
           -------->
Moment-4-Life::Nicki-Minaj
[confidence: 0.511727, lift: 29.157197]

Did-It-On'-Em::Nicki-Minaj
           -------->
Look-At-Me-Now::Chris-Brown
[confidence: 0.528139, lift: 49.902474]

Back-Seat::New-Boyz
           -------->
Moment-4-Life::Nicki-Minaj
[confidence: 0.617512, lift: 35.184585]

We-R-Who-We-R::Ke$ha
           -------->
Grenade::Bruno-Mars
[confidence: 0.514925, lift: 34.843564]

Look-At-Me-Now::Chris-Brown
           -------->
Moment-4-Life::Nicki-Minaj
[confidence: 0.507973, lift: 28.943278]

Look-At-Me-Now::Chris-Brown
           -------->
Down-On-Me::Jeremih
[confidence: 0.535308, lift: 32.701849]

Yeah-3X::Chris-Brown
           -------->
Grenade::Bruno-Mars
[confidence: 0.502538, lift: 34.005349]

Written-In-The-Stars::Tinie-Tempah
           -------->
Down-On-Me::Jeremih
[confidence: 0.535294, lift: 32.701031]

Down-On-Me::Jeremih
           -------->
Moment-4-Life::Nicki-Minaj
[confidence: 0.511046, lift: 29.11837

### metrics: lift

In [35]:
ruleList_lift = ap.generateRules(Lift, support, metric='lift', minMetric=20)

frozenset([34]) --> frozenset([80]) conf: 0.364985163205  lift: 28.4578657326
frozenset([80]) --> frozenset([34]) conf: 0.231203007519  lift: 28.4578657326
frozenset([2791]) --> frozenset([634]) conf: 0.297935103245  lift: 37.9090431981
frozenset([634]) --> frozenset([2791]) conf: 0.30981595092  lift: 37.9090431981
frozenset([3003]) --> frozenset([635]) conf: 0.404255319149  lift: 47.3686741195
frozenset([635]) --> frozenset([3003]) conf: 0.268361581921  lift: 47.3686741195
frozenset([2774]) --> frozenset([20]) conf: 0.440298507463  lift: 27.9259664978
frozenset([20]) --> frozenset([2774]) conf: 0.180428134557  lift: 27.9259664978
frozenset([20]) --> frozenset([18]) conf: 0.227828746177  lift: 27.0009611184
frozenset([18]) --> frozenset([20]) conf: 0.425714285714  lift: 27.0009611184
frozenset([89]) --> frozenset([2840]) conf: 0.308823529412  lift: 27.3134328358
frozenset([2840]) --> frozenset([89]) conf: 0.268656716418  lift: 27.3134328358
frozenset([20]) --> frozenset([34]) conf: 0.1

In [40]:
ap.pntRules(ruleList_lift, item_meaning)

Deuces::Chris-Brown-&-Tyga
           -------->
No-Hands::Waka-Flocka-Flame
[confidence: 0.364985, lift: 28.457866]

No-Hands::Waka-Flocka-Flame
           -------->
Deuces::Chris-Brown-&-Tyga
[confidence: 0.231203, lift: 28.457866]

Colder-Weather::Zac-Brown-Band
           -------->
Who-Are-You-When-I'm-Not-Looking::Blake-Shelton
[confidence: 0.297935, lift: 37.909043]

Who-Are-You-When-I'm-Not-Looking::Blake-Shelton
           -------->
Colder-Weather::Zac-Brown-Band
[confidence: 0.309816, lift: 37.909043]

Bleed-Red::Ronnie-Dunn
           -------->
Are-You-Gonna-Kiss-Me-Or-Not?::Thompson-Square
[confidence: 0.404255, lift: 47.368674]

Are-You-Gonna-Kiss-Me-Or-Not?::Thompson-Square
           -------->
Bleed-Red::Ronnie-Dunn
[confidence: 0.268362, lift: 47.368674]

We-R-Who-We-R::Ke$ha
           -------->
What's-My-Name::Rihanna
[confidence: 0.440299, lift: 27.925966]

What's-My-Name::Rihanna
           -------->
We-R-Who-We-R::Ke$ha
[confidence: 0.180428, lift: 27.925966]

What's

## 2d.

Identify 3-4 rules and explain why you think they are interesting.

__Rule 1:__

Every-Little-Thing::J-Boog
           -------->
Feels-Like::Katchafire
[confidence: 0.728814, lift: 243.799891]

__Conclusion : about 72% of time people, who listen to Every-Little-Thing::J-Boog, listen to Feels-Like::Katchafire__




__Rule 2:__

Grenade::Bruno-Mars
6-Foot-7-Foot::Lil-Wayne
           -------->
No-Hands::Waka-Flocka-Flame
Down-On-Me::Jeremih
[confidence: 0.597315, lift: 104.103547]

__Conclusion : high value of lift indicate that people who like Grenade::Bruno-Mars and 6-Foot-7-Foot::Lil-Wayne also like No-Hands::Waka-Flocka-Flame anc Down-On-Me::Jeremih__





__Rule 3:__

Grenade::Bruno-Mars
6-Foot-7-Foot::Lil-Wayne
           -------->
No-Hands::Waka-Flocka-Flame
Moment-4-Life::Nicki-Minaj
[confidence: 0.624161, lift: 117.150232]

Grenade::Bruno-Mars
Moment-4-Life::Nicki-Minaj
           -------->
No-Hands::Waka-Flocka-Flame
6-Foot-7-Foot::Lil-Wayne
[confidence: 0.531429, lift: 113.627099]

No-Hands::Waka-Flocka-Flame
6-Foot-7-Foot::Lil-Wayne
           -------->
Grenade::Bruno-Mars
Moment-4-Life::Nicki-Minaj
[confidence: 0.479381, lift: 113.627099]

No-Hands::Waka-Flocka-Flame
Moment-4-Life::Nicki-Minaj
           -------->
Grenade::Bruno-Mars
6-Foot-7-Foot::Lil-Wayne
[confidence: 0.420814, lift: 117.150232]


__Conclusion : high value of lift among these 4 songs indicate that people who like two of these songs like to listen other two__



## Q3: Item-Based Joke Recommendation 

[Dataset: jokes.zip]

For this problem you will use a modified version of the item-based recommender algorithm from Ch. 14 of Machine Learning in Action and use it on joke ratings data based on Jester Online Joke Recommender System. The modified version of the code is provided in the module itemBasedRec.py. Most of the module will be used as is, but you will add some additional functionality.

The data set contains two files. The file "modified_jester_data.csv" contains the ratings on 100 jokes by 1000 users (each row is a user profile). The ratings have been normalized to be between 1 and 21 (a 20-point scale), with 1 being the lowest rating. A zero indicated a missing rating. The file "jokes.csv" contains the joke ids mapped to the actual text of the jokes.

Your tasks in this problem are the following (please also see comments for the function stubs in the provided module):

In [ ]:
from numpy import *
from numpy import linalg as la
import numpy as np

def ecludSim(inA,inB):
    return 1.0 / (1.0 + la.norm(inA - inB))

def pearsSim(inA,inB):
    if len(inA) < 3 : return 1.0
    return 0.5 + 0.5 * corrcoef(inA, inB, rowvar = 0)[0][1]

def cosSim(inA,inB):
    num = float(inA.T * inB)
    denom = la.norm(inA)*la.norm(inB)
    return 0.5 + 0.5 * (num / denom)

def standEst(dataMat, user, simMeas, item):
    n = shape(dataMat)[1]
    simTotal = 0.0; ratSimTotal = 0.0
    for j in range(n):
        userRating = dataMat[user,j]
        if userRating == 0: continue
        overLap = nonzero(logical_and(dataMat[:,item]>0, \
                                      dataMat[:,j]>0))[0]
        if len(overLap) == 0: similarity = 0
        else: similarity = simMeas(dataMat[overLap,item], \
                                   dataMat[overLap,j])
        #print 'the %d and %d similarity is: %f' % (item, j, similarity)
        simTotal += similarity
        ratSimTotal += similarity * userRating
    if simTotal == 0: return 0
    else: return ratSimTotal/simTotal
    
def svdEst(dataMat, user, simMeas, item):
    n = shape(dataMat)[1]
    simTotal = 0.0; ratSimTotal = 0.0
    data=mat(dataMat)
    U,Sigma,VT = la.svd(data)
    Sig4 = mat(eye(4)*Sigma[:4]) #arrange Sig4 into a diagonal matrix
    xformedItems = data.T * U[:,:4] * Sig4.I  #create transformed items
    for j in range(n):
        userRating = data[user,j]
        if userRating == 0 or j==item: continue
        similarity = simMeas(xformedItems[item,:].T,\
                             xformedItems[j,:].T)
        #print 'the %d and %d similarity is: %f' % (item, j, similarity)
        simTotal += similarity
        ratSimTotal += similarity * userRating
    if simTotal == 0: return 0
    else: return ratSimTotal/simTotal

# This function is not needed for Assignment 4, but may be useful for experimentation
def recommend(dataMat, user, N=3, simMeas=cosSim, estMethod=standEst):
    unratedItems = nonzero(dataMat[user,:].A==0)[1] #find unrated items 
    if len(unratedItems) == 0: return 'you rated everything'
    itemScores = []
    for item in unratedItems:
        estimatedScore = estMethod(dataMat, user, simMeas, item)
        itemScores.append((item, estimatedScore))
    return sorted(itemScores, key=lambda jj: jj[1], reverse=True)[:N]

# This function performs cross-validation on a single user based on the test_ratio
# For example, with test_ratio = 0.2, 5-fold x-validation is performed where in each fold, 
# 20 percent of rated items are withheld and the rest are used to estimate the withheld ratings

def cross_validate_user(dataMat, user, test_ratio, estMethod=standEst, simMeas=pearsSim):
    number_of_items = np.shape(dataMat)[1]
    rated_items_by_user = np.array([i for i in range(number_of_items) if dataMat[user,i]>0])
    test_size = test_ratio * len(rated_items_by_user)
    test_indices = np.random.randint(0, len(rated_items_by_user), test_size)
    withheld_items = rated_items_by_user[test_indices]
    original_user_profile = np.copy(dataMat[user])
    dataMat[user, withheld_items] = 0 # So that the withheld test items is not used in the rating estimation below
    error_u = 0.0
    count_u = len(withheld_items)

    # Compute absolute error for user u over all test items
    for item in withheld_items:
        # Estimate rating on the withheld item
        estimatedScore = estMethod(dataMat, user, simMeas, item)
        error_u = error_u + abs(estimatedScore - original_user_profile[item])	

    # Now restore ratings of the withheld items to the user profile
    for item in withheld_items:
        dataMat[user, item] = original_user_profile[item]

    # Return sum of absolute errors and the count of test cases for this user
    # Note that these will have to be accumulated for each user to compute MAE
    return error_u, count_u


def load_jokes(file):
    jokes = genfromtxt(file, delimiter=',', dtype=str)
    jokes = np.array(jokes[:,1])
    return jokes

def get_joke_text(jokes, id):
    return jokes[id]

# dataMat = genfromtxt('modified_jester_data.csv',delimiter=',')

# test(dataMat, 0.2, svdEst)
# test(dataMat, 0.2, standEst)

# jokes = load_jokes('jokes.csv')
# print_most_similar_jokes(dataMat, jokes, 3, 5, pearsSim)

''' See example output below:

Selected joke: 

Q. What's the difference between a man and a toilet? A. A toilet doesn't follow you around after you use it.

Top 5 Recommended jokes are :

Q: What's the difference between a Lawyer and a Plumber? A: A Plumber works to unclog the system. 
_______________
What do you call an American in the finals of the world cup? "Hey Beer Man!" 
_______________
Q. What's 200 feet long and has 4 teeth? <P>A. The front row at a Willie Nelson Concert. 
_______________
A country guy goes into a city bar that has a dress code and the maitred' demands he wear a tie. Discouraged the guy goes to his car to sulk when inspiration strikes: He's got jumper cables in the trunk! So he wrapsthem around his neck sort of like a string tie (a bulky string tie to be sure) and returns to the bar. The maitre d' is reluctant but says to the guy "Okay you're a pretty resourceful fellow you can come in... but just don't start anything"!   
_______________
What do you get when you run over a parakeet with a lawnmower? <P>Shredded tweet. 
_______________

'''

## 3a.

Load in the joke ratings data and the joke text data into appropriate data structures.

In [74]:
dataMat = np.genfromtxt('jokes/modified_jester_data.csv',delimiter=',')
jokes = load_jokes('jokes/jokes.csv')

In [75]:
dataMat

array([[  3.18,  19.79,   1.34, ...,   0.  ,   0.  ,   0.  ],
       [ 15.08,  10.71,  17.36, ...,  11.34,   6.68,  12.07],
       [  0.  ,   0.  ,   0.  , ...,   0.  ,   0.  ,   0.  ],
       ..., 
       [ 16.58,  16.63,  15.85, ...,   0.  ,   0.  ,   0.  ],
       [  3.67,   4.45,   3.67, ...,   3.77,   3.77,   3.28],
       [  9.88,  11.73,   9.16, ...,   0.  ,   0.  ,   0.  ]])

In [76]:
jokes[0]

'A man visits the doctor. The doctor says "I have bad news for you.You have cancer and Alzheimer\'s disease". The man replies "Well thank God I don\'t have cancer!"'

## 3b.

Complete the definition for the function "test". This function iterates over all users and for each performs cross-validation on items (by calling the provided "cross_validate_user" function), and returns the error information necessary to compute Mean Absolute Error (MAE). Use this function to perform 5-fold cross-validation (i.e., 20% test-ratio) comparing MAE results using standard item-based collaborative filtering (based on the rating prediction function "standEst") with results using the SVD-based version of the rating item-based CF (using "svdEst" as the prediction engine). [Note: See comments provided in the module for hints on accomplishing these tasks.]

In [62]:
np.shape(dataMat)[0]

1000

In [63]:
def test(dataMat, test_ratio, Method):
    MAE=0
    error = 0
    count = 0
    for i in range(np.shape(dataMat)[0]):
        if Method == "svdEst":
            err, ct = cross_validate_user(dataMat, i, test_ratio, estMethod=svdEst)
        elif Method == "standEst":
            err, ct = cross_validate_user(dataMat, i, test_ratio, estMethod=standEst)
        error += err
        count += ct
    MAE = error/count
    print 'Mean Absoloute Error for', Method, ':', MAE

In [71]:
test(dataMat, 0.2, "svdEst")


/Users/jasonwu/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:67: VisibleDeprecationWarning: using a non-integer number instead of an integer will result in an error in the future


Mean Absoloute Error for svdEst : 3.68118452284


In [72]:
test(dataMat, 0.2, "standEst")

/Users/jasonwu/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:67: VisibleDeprecationWarning: using a non-integer number instead of an integer will result in an error in the future


Mean Absoloute Error for standEst : 3.71676282398


## 3c.

Write a new function "print_most_similar_jokes" which takes the joke ratings data, a query joke id, a parameter k for the number of nearest neighbors, and a similarity metric function, and prints the text of the query joke as well as the texts of the top k most similar jokes based on user ratings. [Note: For hints on how to accomplish this task, please see comments at the end of the provided module as well as comments for the provided stub function.]

In [82]:
def print_most_similar_jokes(dataMat, jokes, queryJoke, k, metric=pearsSim):
    import collections
    query_joke_rating = dataMat.T[queryJoke]
    sim = {}
    index = 0
    for i in dataMat.T:
        similarity = metric(query_joke_rating, i)
        if similarity == 1:
            similarity = 0 # ignore itself
        sim[index] = similarity
        index += 1
    sorted_sim = collections.Counter(sim) 
    print 'Selected joke:'
    print '%s' % jokes[queryJoke], '\n'
    print 'Top', k, 'Recommended jokes are:', '\n'
    for key, value in sorted_sim.most_common(k):
        print '%s' % jokes[key], '\n', '___________________'

In [84]:
print_most_similar_jokes(dataMat, jokes, 3, 5, pearsSim)

Selected joke:
Q. What's the difference between a man and a toilet? A. A toilet doesn't follow you around after you use it. 

Top 5 Recommended jokes are: 

What do you get when you run over a parakeet with a lawnmower? Shredded tweet. 
___________________
A country guy goes into a city bar that has a dress code and the maitred' demands he wear a tie. Discouraged the guy goes to his car to sulk when inspiration strikes: He's got jumper cables in the trunk! So he wraps them around his neck sort of like a string tie (a bulky string tie to be sure) and returns to the bar. The maitre d' is reluctant but says to the guy "Okay you're a pretty resourceful fellow you can come in... but just don't start anything"! 
___________________
Q. What's 200 feet long and has 4 teeth? A. The front row at a Willie Nelson Concert. 
___________________
What do you call an American in the finals of the world cup?"Hey Beer Man!" 
___________________
Q: What's the difference between a Lawyer and a Plumber? A: 